File: backend/core/src/main/java/com/gte619n/healthfitness/core/goals/chat/GoalProposal.java
package com.gte619n.healthfitness.core.goals.chat;

import com.gte619n.healthfitness.core.goals.Comparator;
import com.gte619n.healthfitness.core.goals.GoalDomain;
import com.gte619n.healthfitness.core.goals.StepKind;
import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

/**
 * A validated Goal structure proposed by Gemini (or hand-edited by the
 * user before commit).
 *
 * <p>This is the canonical in-core representation produced by
 * {@link GoalProposalValidator}. Every level carries an optional
 * {@code validationError} string: when validation flags a field, the
 * offending value is <strong>kept</strong> and the error is attached
 * inline so the UI can render it on the proposal card, rather than
 * silently dropping the field (IMPL-12 acceptance criteria).
 *
 * <p>Date fields are normalized to {@link LocalDate}; {@code countFrom}
 * is normalized to {@link Instant} and defaulted to "now" for COUNT
 * steps during validation.
 */
public record GoalProposal(
    String title,
    String description,
    GoalDomain domain,
    LocalDate targetDate,
    List<ProposalPhase> phases,
    // Goal-level error (e.g. targetDate in the past, missing title).
    String validationError
) {

    public record ProposalPhase(
        String title,
        String description,
        LocalDate targetStartDate,
        LocalDate targetEndDate,
        List<ProposalStep> steps,
        String validationError
    ) {}

    public record ProposalStep(
        String title,
        StepKind kind,
        ProposalMetric metric,   // null for MANUAL
        String validationError
    ) {}

    public record ProposalMetric(
        String metricKey,
        Comparator comparator,
        Double targetValue,
        Integer windowDays,   // SUSTAINED only
        Instant countFrom,    // COUNT only
        String validationError
    ) {}

    /**
     * True when no level of the proposal carries a validationError.
     * The commit endpoint refuses to persist a proposal that isn't
     * fully valid.
     */
    public boolean isValid() {
        if (validationError != null) return false;
        if (phases == null) return true;
        for (ProposalPhase p : phases) {
            if (p.validationError() != null) return false;
            if (p.steps() == null) continue;
            for (ProposalStep s : p.steps()) {
                if (s.validationError() != null) return false;
                ProposalMetric m = s.metric();
                if (m != null && m.validationError() != null) return false;
            }
        }
        return true;
    }
}

File: android/core-domain/src/main/java/com/gte619n/healthfitness/domain/goals/GoalProposal.kt
package com.gte619n.healthfitness.domain.goals

// Kotlin twin of the backend GoalProposalDto (see backend
// api/goals/dto/GoalProposalDto.java). Streamed as the SSE `proposal` event's
// JSON and POSTed back (user-edited) to the commit endpoint. Each level carries
// an optional `validationError` so the editable card can flag offending fields
// inline rather than dropping them (IMPL-12 acceptance criteria).
//
// Dates are LocalDate-style ISO strings ("2026-09-30"); countFrom is an Instant
// ISO string. Kept as strings here for the same reason as Goals.kt — parsed in
// the UI layer (assumption 18).

data class GoalProposal(
    val title: String? = null,
    val description: String? = null,
    val domain: GoalDomain? = null,
    val targetDate: String? = null,
    val phases: List<ProposalPhase> = emptyList(),
    val validationError: String? = null,
)

data class ProposalPhase(
    val title: String? = null,
    val description: String? = null,
    val targetStartDate: String? = null,
    val targetEndDate: String? = null,
    val steps: List<ProposalStep> = emptyList(),
    val validationError: String? = null,
)

data class ProposalStep(
    val title: String? = null,
    val kind: StepKind = StepKind.MANUAL,
    val metric: ProposalMetric? = null,
    val validationError: String? = null,
)

data class ProposalMetric(
    val metricKey: String? = null,
    val comparator: Comparator? = null,
    val targetValue: Double? = null,
    val windowDays: Int? = null,
    val countFrom: String? = null,
    val validationError: String? = null,
)

/** Response body of POST .../chat/{threadId}/commit on success. */
data class CommitResponse(val goalId: String)

/**
 * The fixed metric-key registry the backend resolver understands (IMPL-12).
 * Used by the proposal-card metric editor's dropdown so the user can only bind
 * Steps to real metrics. Mirrors the spec's registry verbatim.
 */
object MetricRegistry {
    val keys: List<String> = listOf(
        "body.weight",
        "body.bodyFatPct",
        "body.leanMass",
        "blood.ldl",
        "blood.apoB",
        "blood.hba1c",
        "blood.hsCRP",
        "vitals.restingHr",
        "vitals.hrv",
        "vitals.sleepScore",
        "workouts.count",
        "workouts.weeklyVolume",
        "nutrition.proteinAvg7d",
        "nutrition.carbsAvg7d",
        "nutrition.fatAvg7d",
        "nutrition.caloriesAvg7d",
        "meds.adherence30d",
    )
}
